include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
-DEFINE_FIND_MSB_COMMON_SPECIALIZATION(uint8_t, uint32_t)
-#endif
+template<typename T> AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(clamp_helper, clamp, concepts::Scalar<T>, (T), (T)(T)(T), T)
+template<typename T> AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(max_helper, max, concepts::Scalar<T>, (T), (T)(T), T)
+template<typename T> AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(min_helper, min, concepts::Scalar<T>, (T), (T)(T), T)
+
+#undef DECL_ARG
+#undef WRAP
+#undef ARG
+#undef AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER
 
-template<>
-struct find_msb_helper<uint64_t>
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct bitReverse_helper<T>
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(uint64_t) val)
+	static inline T __call(NBL_CONST_REF_ARG(T) arg)
 	{
-#ifdef __HLSL_VERSION
-		const uint32_t highBits = uint32_t(val >> 32);
-		const int32_t highMsb = find_msb_helper<uint32_t>::findMSB(highBits);
-
-		if (highMsb == -1)
-		{
-			const uint32_t lowBits = uint32_t(val);
-			const int32_t lowMsb = find_msb_helper<uint32_t>::findMSB(lowBits);
-			if (lowMsb == -1)
-				return -1;
-
-			return lowMsb;
-		}
-
-		return highMsb + 32;
-#else
-		return glm::findMSB(val);
-#endif
+		return glm::bitfieldReverse<T>(arg);
 	}
 };
+template<typename Matrix>
+requires concepts::Matrix<Matrix>
+struct transpose_helper<Matrix>
+{
+	using transposed_t = typename matrix_traits<Matrix>::transposed_type;
 
-template<int N>
-struct find_msb_helper<vector<uint32_t, N> >
+	static transposed_t __call(NBL_CONST_REF_ARG(Matrix) m)
+	{
+		return reinterpret_cast<transposed_t&>(glm::transpose(reinterpret_cast<typename Matrix::Base const&>(m)));
+	}
+};
+template<typename Vector>
+requires concepts::FloatingPointVector<Vector>
+struct length_helper<Vector>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	static inline typename vector_traits<Vector>::scalar_type __call(NBL_CONST_REF_ARG(Vector) vec)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findUMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return std::sqrt(dot_helper<Vector>::__call(vec, vec));
 	}
 };
-
-template<int N>
-struct find_msb_helper<vector<int32_t, N> >
+template<typename Vectorial>
+requires concepts::FloatingPointLikeVectorial<Vectorial>
+struct normalize_helper<Vectorial>
 {
-	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	static inline Vectorial __call(NBL_CONST_REF_ARG(Vectorial) vec)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findSMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return vec / length_helper<Vectorial>::__call(vec);
 	}
 };
 
-#ifndef __HLSL_VERSION
-
+template<typename T>
+requires concepts::IntegralScalar<T>
+struct find_lsb_helper<T>
+{
+	using return_t = int32_t;
+	static inline T __call(const T arg)
+	{
+		return glm::findLSB<T>(arg);
+	}
+};
+template<typename Integer>
+NBL_PARTIAL_REQ_TOP(concepts::IntegralScalar<Integer>)
+struct find_msb_helper<Integer NBL_PARTIAL_REQ_BOT(concepts::IntegralScalar<Integer>) >
```

Comment:
if you're using `requires` and c++20 syntax in the c++ section then keep doing that, be consistent